modules/core/src/opencl/intel_gemm.cl

```diff
@@ -0,0 +1,1080 @@
+    vstore4(dot07, 0, dst_write0); dst_write0 += ldC;
+}
+
+#undef VEC_SIZE
+#undef LWG_HEIGHT
+#undef TILE_M
+#undef TILE_K
+#undef TILE_N
+
+#define VEC_SIZE        4
+#define LWG_HEIGHT      4
+#define TILE_M          8
+#define TILE_K          16
+#define TILE_N          32
+
+__attribute__((reqd_work_group_size(8, LWG_HEIGHT, 1)))
+__kernel void intelblas_gemm_buffer_NN(
+    const __global float *src0, int off0,
+    const __global float *src1, int off1,
+    __global float *dst, int offd,
+    int M,
+    int N,
+    int K,
+    float alpha,
+    float beta,
+    int ldA,
+    int ldB,
+    int ldC,
+    int start_index,
+    int stride)
+{
+    const int group_x = get_group_id(0);
+    const int group_y = get_group_id(1);
+    const int local_x = get_local_id(0);
+    const int local_y = get_local_id(1);
+    const int global_x = get_global_id(0);
+    const int global_y = get_global_id(1);
+
+    float4 brow;
+    float2 arow0, arow1, arow2, arow3, arow4, arow5, arow6, arow7;
+
+    __global float *dst_write0 = dst + local_x * VEC_SIZE + ( group_x * TILE_N ) + ( group_y * LWG_HEIGHT * TILE_M + local_y * TILE_M) * ldC + offd;
+
+    const __global float *src0_read = src0 + local_x * ( TILE_K / 8 ) + ( group_y * LWG_HEIGHT * TILE_M + local_y * TILE_M ) * ldA + start_index + off0;
+
+    const __global float *src1_read0 = src1 + local_x * VEC_SIZE + ( group_x * TILE_N ) + start_index * ldB + off1;
+
+    int border = -(group_y * LWG_HEIGHT * TILE_M + local_y * TILE_M);
+
+    int row0 = mad24(global_y, TILE_M, 0) < M ? 0 : border;
+    int row1 = mad24(global_y, TILE_M, 1) < M ? 1 : border;
+    int row2 = mad24(global_y, TILE_M, 2) < M ? 2 : border;
+    int row3 = mad24(global_y, TILE_M, 3) < M ? 3 : border;
+    int row4 = mad24(global_y, TILE_M, 4) < M ? 4 : border;
+    int row5 = mad24(global_y, TILE_M, 5) < M ? 5 : border;
+    int row6 = mad24(global_y, TILE_M, 6) < M ? 6 : border;
+    int row7 = mad24(global_y, TILE_M, 7) < M ? 7 : border;
+
+    float4 dot00 = (start_index != 0) ? vload4(0, dst_write0) : (float)beta * vload4(0, dst_write0);
+    float4 dot01 = (start_index != 0) ? vload4(0, dst_write0 + 1 * ldC) : (float)beta * vload4(0, dst_write0 + 1 * ldC);
+    float4 dot02 = (start_index != 0) ? vload4(0, dst_write0 + 2 * ldC) : (float)beta * vload4(0, dst_write0 + 2 * ldC);
+    float4 dot03 = (start_index != 0) ? vload4(0, dst_write0 + 3 * ldC) : (float)beta * vload4(0, dst_write0 + 3 * ldC);
+    float4 dot04 = (start_index != 0) ? vload4(0, dst_write0 + 4 * ldC) : (float)beta * vload4(0, dst_write0 + 4 * ldC);
+    float4 dot05 = (start_index != 0) ? vload4(0, dst_write0 + 5 * ldC) : (float)beta * vload4(0, dst_write0 + 5 * ldC);
+    float4 dot06 = (start_index != 0) ? vload4(0, dst_write0 + 6 * ldC) : (float)beta * vload4(0, dst_write0 + 6 * ldC);
+    float4 dot07 = (start_index != 0) ? vload4(0, dst_write0 + 7 * ldC) : (float)beta * vload4(0, dst_write0 + 7 * ldC);
```

Comment:
This vload4 series probably reads out of buffer.
Including nan values:

Check code:
```
#define CHECK_NAN_(id, v) if(isnan(dot ## id . s ## v)) { printf("dot" #id ".s" #v " is NAN, lx=%d ly=%d\n", local_x, local_y); }
#define CHECK_NAN(id) CHECK_NAN_(id, 0) CHECK_NAN_(id, 1) CHECK_NAN_(id, 2) CHECK_NAN_(id, 3)
    CHECK_NAN(00) CHECK_NAN(01) CHECK_NAN(02) CHECK_NAN(03)
    CHECK_NAN(04) CHECK_NAN(05) CHECK_NAN(06) CHECK_NAN(07)
```

Sporadic results (dst in this case rows=10 cols=5):
```
OpenCV: OCL: Kernel::run(intelblas_gemm_buffer_NN)
    dims=2 local=8x4x1 global=8x4x1
dot00.s0 is NAN, lx=3 ly=3
dot00.s0 is NAN, lx=1 ly=2
dot00.s0 is NAN, lx=7 ly=2
dot00.s0 is NAN, lx=5 ly=1
dot01.s3 is NAN, lx=1 ly=3
...
```

Probably these nan values are propagated via pipeline. I hope it is the reason of nan values from this [comment](https://github.com/opencv/opencv/pull/8104#issuecomment-286280174).